cirq/circuits/circuit.py

```diff
@@ -1258,6 +1258,109 @@ def _json_dict_(self):
+
+            a) Just before their operations would collide.
+            b) stop_at_first_alignment==True and their ends (or starts) align
+                for the first time.
+            c) Their ends (or starts) align for the last time.
+
+        Beware that this method is *not* associative. For example:
+
+            >>> a, b = cirq.LineQubit.range(2)
+            >>> A = cirq.Circuit(cirq.H(a))
+            >>> B = cirq.Circuit(cirq.H(b))
+            >>> f = cirq.Circuit.tetris_concat
+            >>> f(f(A, B), A) == f(A, f(B, A))
+            False
+            >>> len(f(f(f(A, B), A), B)) == len(f(f(A, f(B, A)), B))
+            False
+
+        Args:
+            circuits: The circuits to concatenate.
+            stop_at_first_alignment: Defaults to false. When true, the circuits
+                are never overlapped more than needed to align their starts or
+                to align their ends. When false, the smaller circuit can be
+                pushed deeper into the larger circuit, past the first time their
+                starts or ends align, until the second time their starts or ends
+                align.
+
+        Returns:
+            The concatenated and overlapped circuit.
+        """
+        if len(circuits) == 0:
+            return Circuit()
+        n_acc = len(circuits[0])
+
+        # Allocate a buffer large enough to append and prepend all the circuits.
+        pad_len = sum(len(c) for c in circuits) - n_acc
+        buffer = np.zeros(shape=pad_len * 2 + n_acc, dtype=np.object)
+
+        # Put the initial circuit in the center of the buffer.
+        offset = pad_len
+        buffer[offset : offset + n_acc] = circuits[0].moments
+
+        # Accumulate all the circuits into the buffer.
+        for k in range(1, len(circuits)):
+            offset, n_acc = _tetris_concat_helper(
+                offset, n_acc, buffer, circuits[k].moments, stop_at_first_alignment
+            )
+
+        return cirq.Circuit(buffer[offset : offset + n_acc])
+
+
+def _overlap_collision_time(
+    c1: Sequence['cirq.Moment'], c2: Sequence['cirq.Moment'], stop_at_first_alignment: bool
+) -> int:
+    # Tracks the first used moment index for each qubit in c2.
+    # Tracks the complementary last used moment index for each qubit in c1.
+    seen_times: Dict['cirq.Qid', int] = {}
+
+    # Start scanning from end of first and start of second.
+    upper_bound = (min if stop_at_first_alignment else max)(len(c1), len(c2))
+    t = 0
+    while t < upper_bound:
+        if t < len(c2):
+            for op in c2[t]:
+                for q in op.qubits:
+                    # Record time but check if qubit already seen on other side.
+                    k2 = seen_times.setdefault(q, t)
+                    if k2 < 0:
+                        # Use this qubit collision to bound the collision time.
+                        upper_bound = min(upper_bound, t + ~k2)
```

Comment:
Using int inversion is a neat trick, but could use a bit of clarification comment maybe above when you explain the "complementary last used index in c1".

Reply:
That comment is at the start of the method.

Reply:
I added a comment noting the complement.